### ISSUE TITLE
Device/Mobile Tracking

### ISSUE THREAD

**Issue by Quin452** [offen/offen#724]
It says in your Docs that you don't use user agent string for tracking, but instead runs a check to see if the device can rotate.
However, I have just tested this on my mobile, and I cannot see any data reflected on the analytics.

Is this scrapped?

Could there be a way of tracking this data, without invading privacy?  As a web developer, I'm thinking something along the lines of screen-size or orientation (which, would actually be useful for me as a developer for responsiveness).

**Comment by m90** [comment]
There is a comment in the source code about this likely working less better in the future it seems: https://github.com/offen/offen/blob/369e2c991400ea1d0175b0f80fe4796fcb46b92a/script/src/events.js#L23-L25

so it's not too unexpected.

If we wanted to keep this simple and privacy friendly by just signaling "Mobile: yes/no" screen resolution won't work too well as today's high end phones have a higher resolution than the laptop I am typing this on. I'm not sure which other means of detecting this there are out there, but in case you or someone else knows of a good way to do so I'm definitely open to changing how this detect works.

**Comment by Quin452** [comment]
Thanks for the information.

You are true that mobile devices are becoming more high-scale.  I'd recommend checking if the screen is touch-enabled, although there are more devices with touchscreens nowadays (I've got a [dead] laptop that has touchscreen).

_This actually poses an issue with responsive web design (I'll worry about that later), so I'll need to do some testing to see what happens with media queries._

Could getting the device's viewport be useful?  A few lines of code to get the device/window width and height?
It wouldn't be as simple as bool, but it still would be useful data.  You could extrapolate if the user is viewing the "mobile" design or the "desktop" design.

**Comment by m90** [comment]
> I'd recommend checking if the screen is touch-enabled, although there are more devices with touchscreens nowadays

Yeah, touch unfortunately has too many false positives caused by laptops with a Touchscreen so that'd be just as inaccurate as the current approach.

> Could getting the device's viewport be useful? A few lines of code to get the device/window width and height?

I'm not sure if we want to start collecting the actual viewport size and start displaying it in the stats (correct me if I'm wrong @hendr-ik). Having the boolean "Mobile" signal seems to be a good compromise between privacy and usefulness assuming it works well.

Question is: how do we get there? It'd be nice if we could find a solution that does work without having to look at the UA string, but that's maybe not a 100% hard blocker.

Considering the current approach will work each day in the future while the API is being deprecated we'll have to come up with some solution anyways. We'll add this to our agenda. Any ideas about how to implement this are welcome.

**Comment by Quin452** [comment]
Maybe instead of displaying a bunch of resolution stats (again, it would be extremely helpful as a web developer, so maybe a sneaky hidden area for me :-P) you could use their viewports to categorise them as "desktop" or "mobile".

This wouldn't be based on their device, but literally how they are viewing the site; so people viewing the site with a resized window (and thus, the responsive design) would be "mobile".

Not 100% accurate, but it would show how the site is being viewed.

**Comment by m90** [comment]
I'm not entirely sure I understand what you mean when you say "viewport". Is this just the size of the browser window or does it include other factors as well?

As said earlier, my laptop's resolution (1920x1080) is smaller than a newish Samsung Galaxy (2340x1080), so I'm not sure if this allows a better quality than the current approach. I start wondering if the screen's aspect ration could be a good hint though.

**Comment by Quin452** [comment]
Viewport is different from resolution.  They may match when in fullscreen (F11), but with the URL bar, Bookmarks, etc., you may maximise the window, but it won't be 1920x1080.  The viewport would 1920x927px (in my instance).

My phone has a resolution of 2560x1440 on a 5.7" screen, but the actual screensize (including address bar)/viewport (just what is visible) is 408x725 / 408x598 in portrait mode.

You can check for yourself at [What is my viewport](https://whatismyviewport.com/) (I can't post the screenshot).

It really depends on what you'd like, whether you was to get the screen size or the viewport (both can be done with vanilla JS).
Using that, you'd get the viewable area, and then have to do some decision making `if([innerWidth or aspectRatio] < breakpoint) { mobile = true }`

**Comment by m90** [comment]
> My phone has a resolution of 2560x1440 on a 5.7" screen, but the actual screensize (including address bar)/viewport (just what is visible) is 408x725 / 408x598 in portrait mode.

Do you know how this would behave on a laptop that has a so-called "Retina" screen? Would it also divide the measures? I don't have any device with me that I could use for testing this and I'm still a little confused about what's going on?

**Comment by Quin452** [comment]
If we take the Macbook Pro (13" 2018 and 16" 2019 models), both have retina.
They both can have a resolutions up to 2560 x 1600 and 3072 x 1920, but their physical screen sizes limit their viewports to 1280 x 800 and 1536 x 960.  This is why my phone, although can rival some PCs, will always show mobile-friendly sites, but also has the option to "view desktop mode".
And some monitors just cannot handle the larger resolutions; it is why I don't get laptops less than 14" (the resolutions were always too small).

I would liken this to playing games in "fullscreen".  Your computer may have a resolution of 1920x1080, but you set the game resolution to 1024x768 (why would you do that, you monster).
You can see that the game is "scaled down" to fit the game resolution and everything looks bigger.  If you were to set it to 1920x1080, everything would look 1:1.

Or in a similar way, it's like when you plug a new monitor into a Windows 10 machine.  It may set the resolution to the "max size", but it also increases the scale and layout.

I've got 3 monitors, all at 1920x1080, but one of them is physically smaller than the others (by a significant amount).
On this monitor, things just look sharper and clearer, but it also cost far more than the other two combined.

So, even though your phone may be advertised as "4K Resolution", that doesn't mean you can view a site in "desktop mode", because the text and images will be far too small.
The viewport is the actual resolution which is used for responsiveness (which is determined by the advertised resolution, with the physical screen size, to get the pixel per inch).

Essentially put, if you get the window's width and height, or the viewport, you can determine what "view" the user is using.

If you can access Dev Tools (F12 on any browser) you should have an option to view the site as different devices (or use Blisk).  This gives an accurate representation of what mobile users would see, and you'll be able to see what "resolution" they are using.

https://yesviz.com/devices/macbookpro-2018-13/
https://yesviz.com/devices/macbookpro-16-2019/
